python/python/bystro/tada_mixture/mixture_tada_em.py

```diff
@@ -374,6 +427,101 @@ def predict_logproba(self, data):
                 log_proba[i, k] = np.sum(probs)
         return log_proba
 
+    def _store_instance_variables(self, trainable_variables):
+        """
+        Saves the learned variables
+
+        Parameters
+        ----------
+        trainable_variables : list
+            List of variables to save
+        """
+        self.pi = trainable_variables[0]
+        Lambda_list_l = trainable_variables[1]
+        alpha_ls = trainable_variables[2]
+        sigmoid = nn.Sigmoid()
+        self.Lambda = np.zeros((self.K,self.p))
+        self.Alpha = np.zeros((self.K,self.p))
+        for k in range(self.K):
+            Lambda_k = Lambda_list_l[k]
+            alpha_k = sigmoid(alpha_ls[k])
+
+            self.Lambda[k] = Lambda_k.detach().numpy()
+            self.Alpha[k] = alpha_k.detach().numpy()
+
+    def _initialize_variables(self, X):
+        """
+
+        Parameters
+        ----------
+
+        Returns
+        -------
+        """
+        model = GaussianMixture(self.K)
+        model.fit(X)
+        Lambda_list_l = []
+        for i in range(self.K):
+            Lambda_list_l.append(
+                torch.tensor(
+                    1.2 * model.means_[i].astype(np.float32) + 0.1,
+                    requires_grad=True,
+                )
+            )
+
+        pct_0 = np.mean(X == 0, axis=0)
+        pct_0 = np.clip(pct_0, 0.01, 0.99)
+        pct_0_inv = np.log(pct_0 / (1 - pct_0))
+
+        alpha_ls = [
+            torch.tensor(pct_0_inv, requires_grad=True) for i in range(self.K)
+        ]
+        pi_ = 1 / self.K * np.ones(self.K)
+        return Lambda_list_l, alpha_ls, pi_
+
+    def _initialize_save_losses(self):
```

Comment:
it might be difficult for someone walking into this code for the first time to tell what's intended here, because we have a private method whose only content is its docstring, which apparently disagrees with the method itself.  So a reader might wonder: were these methods supposed to be implemented, or are they meant to be this way?  Are they load-bearing at all?  Could I just delete them?

If we just forgot to implement them, let's implement them and put them under test to avoid future confusion.  If they're supposed to be this way, we should probably delete them.  If we can't delete them (maybe because they're required by an ABC further up the class hierarchy?), we should probably reevaluate the class design.

Reply:
I was hoping to save out the ELBO but computing that is not trivial. I was planning on adding it in later, or at least some kind of metric on convergence progress.